### ISSUE TITLE
Feedback-api deploying and integrating in CI/CD

### ISSUE THREAD

**Issue by donpui** [LeastAuthority/feedback-api#5]
We need to create simple CI/CD process and try to deploy container at least to stage/dev server in order to test if we can send emails to feedback@winden.app

Dockerfile image PR https://github.com/LeastAuthority/feedback-api/pull/4

Additionally we can create small verification test:
- launch feedback-api docker image
- launch maildev server (https://github.com/maildev/maildev)
- Send small post message: curl --request POST --header "Content-Type: application/json" --data '{"q1": "a1", "q2": "a2"}' localhost:8001/v1/feedback
- Check if the related mail has been send

Tasks:
- [x] Add Docker Compose support to run both the API and a simple SMTP daemon
- [ ] Add a Github action to validate and test the API using those Docker containers

**Comment by btlogy** [comment]
As discussed in LeastAuthority/it-ops#68 and this morning with @donpui: next step is to deploy the feedback-api in staging so it can be used for testing from winden.

Short list of sub-tasks to consider:
- Publish an image of feedback-api (in Dockerhub, from this repository or from [magic-wormhole-docker](https://github.com/LeastAuthority/magic-wormhole-docker) ?);
- Rework [mw-projects](https://github.com/LeastAuthority/mw-projects) to host additional containers running the feedback-api image (DNS, TLS certificates, Trafiek and Grafana);
- Consider adding some mitigations against DoS or other possible abuses (from [Trafiek](https://doc.traefik.io/traefik/middlewares/http/ratelimit/#ratelimit)?)

**IMPORTANT**: this approach will make it difficult to automate the deployment of feedback-api for 2 reasons:
1. Just like for mailbox and transit images, it will require to commit and merge some version change in this [mw-projetcs](https://github.com/LeastAuthority/mw-projects) repository;
2. Pending on LeastAuthority/mw-projects#73, the deployment of mw backend still has to be manually triggered from a workstation.

**Comment by wuan** [comment]
> Publish an image of feedback-api (in Dockerhub, from this repository or from [magic-wormhole-docker](https://github.com/LeastAuthority/magic-wormhole-docker) ?)
The decision behind having a separate repo for creating containers has been driven from the fact that we do not control the service repositories. In the case of feedback-api we could build images from the repo itself.

**Comment by btlogy** [comment]
> > Publish an image of feedback-api (in Dockerhub, from this repository or from [magic-wormhole-docker](https://github.com/LeastAuthority/magic-wormhole-docker) ?)
> 
> The decision behind having a separate repo for creating containers has been driven from the fact that we do not control the service repositories. In the case of feedback-api we could build images from the repo itself.

Which means we need a DOCKERHUB_USERNAME + DOCKERHUB_TOKEN here too I guess.
As I do not have the proper access to Dockerhub or to the secrets of this GitHub repository:
@wuan, can you take care of this secrets so we can start using them from the actions in the next PR?

**Comment by wuan** [comment]
> Which means we need a DOCKERHUB_USERNAME + DOCKERHUB_TOKEN here too I guess. As I do not have the proper access to Dockerhub or to the secrets of this GitHub repository: @wuan, can you take care of this secrets so we can start using them from the actions in the next PR?

Those variables are now available in this repo.

**Comment by btlogy** [comment]
Re-opening: #11 did not close this issue yet!
I keep seeing issues closed automatically by github bot and I don;t not get it...

There is still work to do here.
For a start: the image is not published as expected: I need to fix the workflow.
And we still not to deliver the app via mw-projects.

**Comment by btlogy** [comment]
So, the image has not been published while merging #11 because of the skip duplicate action:
```
Run fkirc/skip-duplicate-actions@v5
  with:
    concurrent_skipping: same_content_newer
    paths: ["*.go", "go.*", "Dockerfile", ".github/workflows/docker-image.yml"]
    github_token: ***
    paths_ignore: []
    cancel_others: false
    skip_after_successful_duplicate: true
    do_not_skip: ["workflow_dispatch", "schedule"]
Skip execution because the exact same files have been successfully checked in run
```
Which means my logic was flawed: the builds do not have to run in parallel to be considered as duplicates!
Thus, it make sense to force the publication from the main branch regardless: that is what #14 should achieve.

**Comment by btlogy** [comment]
It looks like PR#14 will be blocked by 8ab0346c00ba0d288da9d0f894d2289241c854bd because of a failing check:
```
[/github/workspace/smtp-client.go:76] - G402 (CWE-295): TLS InsecureSkipVerify set true. (Confidence: HIGH, Severity: HIGH)
    75: 		tlsconfig := &tls.Config{
  > 76: 			InsecureSkipVerify: useInsecureTls,
    77: 			ServerName:         hostname,
```
Which is related to our CI requirements!
@donpui : any idea how we can solve that? Should we whitelist? Something like `gosec -exclude=G402`?

**Comment by donpui** [comment]
> any idea how we can solve that? Should we whitelist? Something like `gosec -exclude=G402`?

We could set env: SMTP_USE_TLS=true or SMTP_USE_INSECURE_TLS=false

**Comment by btlogy** [comment]
> We could set env: SMTP_USE_TLS=true or SMTP_USE_INSECURE_TLS=false

Thank you to remind me thos vars. @donpui.
But now, I'm a bit confused about the original reason behind this logic there:
```go
	if useTls || useInsecureTls {
		tlsconfig := &tls.Config{
			InsecureSkipVerify: useInsecureTls,
			ServerName:         hostname,
		}
...
```
When do we need `InsecureSkipVerify=true` at all if we can use `useTls=false` for test and CI?

**Comment by donpui** [comment]
> When do we need `InsecureSkipVerify=true` at all if we can use `useTls=false` for test and CI?

Here are 3 configuration options: 
1. connect to tcp and do not use tls
2. use tls and allow insecure tls usage (with self-sign certificate)
3. use tls and verify certificate

**Comment by btlogy** [comment]
> We could set env: SMTP_USE_TLS=true or SMTP_USE_INSECURE_TLS=false

I might be wrong @donpui, but I don't think changing those var. will impact the gosec action!

**Comment by btlogy** [comment]
> 2\. use tls and allow insecure tls usage (with self-sign certificate)

That was my point @donpui: do we need to use a self-sign certificate if we can just avoid TLS for testing?

**Comment by donpui** [comment]
> I might be wrong @donpui, but I don't think changing those var. will impact the gosec action!

Lets see or we can use ignore statement.

**Comment by btlogy** [comment]
Meanwhile, I've tried to use [annotation](https://github.com/securego/gosec#annotating-code) w/o much [success](https://github.com/LeastAuthority/feedback-api/actions/runs/3996354598/jobs/6856313971):
```

**Comment by donpui** [comment]
> Meanwhile, I've tried to use [annotation](https://github.com/securego/gosec#annotating-code) w/o much [success](https://github.com/LeastAuthority/feedback-api/actions/runs/3996354598/jobs/6856313971):
> 
> ```
> [/github/workspace/smtp-client.go:77] - G402 (CWE-295): TLS InsecureSkipVerify set true. (Confidence: HIGH, Severity: HIGH)
>     76: 			//#nosec G402 -- This is a requirements for CI
>   > 77: 			InsecureSkipVerify: useInsecureTls,
>     78: 			ServerName:         hostname,
> ```

Now solved and pass all checks. Moved ignore statement a little bit above.

**Comment by btlogy** [comment]
Thank you @donpui 83391a7f6bcc4522ae47567ed456fb5bb10fb3da works...
Now should we keep that, I'm not sure, but it will have to be part of some other issue.

I'll merge this new PR with those 2 fixes to hopefully publish to DockerHub...

**Comment by btlogy** [comment]
Again closed automatically: I need to investigate how to prevent this!
However: the delivery in DockerHub succeeded: https://hub.docker.com/r/leastauthority/feedback-http-server/tags

**Comment by btlogy** [comment]
Collecting some 'feedback' from deploying this component:
1. By default, Go smtp [client](https://pkg.go.dev/net/smtp#Client.Hello) initialize the SMTP connection with `HELO localhost`, which can be rejected by some MTA (not sure yet what Google will do):
```
handling a post request to feedback url
sending email via smtp.example.com:25 to {Feedback Mailbox feedback@example.com}
Failed sending feedback, error: 504 5.5.2 <localhost>: Helo command rejected: need fully-qualified hostname
```
2. The current code warns about `SMTP_USERNAME` being not set, even if [authentication](https://pkg.go.dev/net/smtp#PlainAuth) is not required! This could be a lot of useless messages. Maybe we could skip the authentication when the username is not set.
```
	if username == "" {
		log.Printf("WARNING: SMTP_USERNAME not set")
	}
	hostPortStr := fmt.Sprintf("%s:%s", hostname, strconv.Itoa(int(port)))
	auth := smtp.PlainAuth("", username, password, hostname)
```
3. The current code does not seem to support [StartTLS](https://pkg.go.dev/net/smtp#Client.StartTLS). I'm not yet sure it could be a problem with Google, but it is when testing my own MTA...

**Comment by btlogy** [comment]
Something else I've experienced: the current code does not handle SMTP timeout very well:
```
Starting: Tue 14 Feb 2023 10:40:19 AM CET
> POST /v1/feedback HTTP/1.1
> Host: feedback.local
> User-Agent: curl/7.74.0
> Accept: */*
> Content-Type: application/json
> Content-Length: 164
feedback: handling a post request to feedback url
...
...
...
...
feedback: Failed sending feedback, error: dial tcp x.x.x.x:25: connect: connection timed out
traefik: "POST /v1/feedback HTTP/1.1" 502 11 "-" "-" 15 "feedback@docker" "http://172.21.0.5:8001" 129521ms
< HTTP/1.1 502 Bad Gateway
< Date: Tue, 14 Feb 2023 09:42:28 GMT
< Content-Length: 11
< Content-Type: text/plain; charset=utf-8
< Bad Gateway
Stopping: Tue 14 Feb 2023 10:42:28 AM CET
```
There is a 120 seconds timeout on our server side which could lead to bad user experience on the client side.

**Comment by wuan** [comment]
> 1. By default, Go smtp [client](https://pkg.go.dev/net/smtp#Client.Hello) initialize the SMTP connection with `HELO localhost`, which can be rejected by some MTA (not sure yet what Google will do):

I learned that Gmail is quite picky about that.

> Maybe we could skip the authentication when the username is not set.

It would be great to skip authentication if those values are empty.

**Comment by btlogy** [comment]
Now the feedback services has been (partly) deployed on staging, we can see something else is missing to have Google accepting our emails:
```
2023/02/15 15:39:38 sending email via aspmx.l.google.com:25 to {Feedback Mailbox feedback@winden.app}
2023/02/15 15:39:39 Failed sending feedback, error: 550 5.7.1 [51.91.101.169] Messages missing a valid messageId header are not accepted. l5-20020a5d4bc5000000b002c5657dd82fsi5265696wrt.3 - gsmtp
```
This might lead us to speed up the deployment of a proper MTA of our own, because I'm not sure the SMTP Go client will cover this requirement easily.

Any thoughts about that @JustusFT, @vu3rdd or @wuan?

**Comment by btlogy** [comment]
The feedback api has been deployed: see LeastAuhtority/mw-projects#83 for more details.
As it turned out, #20 is not required (yet).
But we better check the headers of the received message before closing this issue...

**Comment by wuan** [comment]
So far this looks good on the receiving end:
```
Received-SPF: pass (google.com: domain of no-reply@mw.leastauthority.com designates 152.228.133.151 as permitted sender) client-ip=152.228.133.151;
Authentication-Results: mx.google.com;
spf=pass (google.com: domain of no-reply@mw.leastauthority.com designates 152.228.133.151 as permitted sender) smtp.mailfrom=no-reply@mw.leastauthority.com
Date: Wed, 15 Feb 2023 14:41:56 -0800 (PST)
Message-ID: <63ed5fb4.050a0220.3d635.109bSMTPIN_ADDED_MISSING@mx.google.com>
From: "Feedback Server" <no-reply@mw.leastauthority.com>
To: "Feedback Mailbox" <feedback@winden.app>
```

Having that "localhost" does not hurt us for now, as Google requires to set up SPF for the sending domain and add the server IPs to the whitelist.

```
Received: from localhost (vps-6b76d505.vps.ovh.net. [152.228.133.151])
by mx.google.com with UTF8SMTP id d5-20020a056402078500b004ab4183a9b7si94202edy.396.2023.02.15.14.41.56
```
